jerry-util/arg-validator/arg-validator.h

```diff
@@ -0,0 +1,94 @@
+ * limitations under the License.
+ */
+
+#ifndef JERRY_ARG_VALIDATOR_H
+#define JERRY_ARG_VALIDATOR_H
+
+#include <assert.h>
+#include <stdbool.h>
+#include <stddef.h>
+#include <stdint.h>
+#include "jerryscript.h"
+
+typedef struct jerry_arg_t jerry_arg_t;
+typedef struct jerry_arg_iterator_t jerry_arg_iterator_t;
+
+/**
+ * Signature of the validator funtion.
+ */
+typedef jerry_value_t (*jerry_arg_func_t) (jerry_arg_iterator_t *iterator_p);
+
+/**
+ * The structure of the jerry argument in validator.
+ * Each native argument should have one corresponding jerry_arg_t item.
+ */
+struct jerry_arg_t
+{
+    jerry_arg_func_t func; /**< the validator function for the native argument */
+    void *dest; /**< point to the native argument */
+    /** The extra infomation for validator */
+    union
+    {
+        void *obj_p; /**< the pointer to native handle's expected type info */
+        jerry_length_t length; /**< the length of the string */
+    } info;
+};
+
+/**
+ * The argument iteration used for validator funtions.
+ */
+struct jerry_arg_iterator_t
+{
+    const jerry_arg_t *c_arg_p; /**< points to the current jerry_arg_t */
+    const jerry_value_t *js_arg_p; /**< points to the current js argument */
+    const jerry_length_t  js_arg_cnt; /**< the total count of the js arguments */
+    jerry_length_t  js_arg_index; /**< the current index of the js argument */
+};
+
+jerry_value_t jerry_validate_and_assign_all (const jerry_value_t this_val,
+                                             const jerry_value_t *js_arg_p,
+                                             const jerry_length_t js_arg_cnt,
+                                             const jerry_arg_t *c_arg_p,
+                                             const jerry_length_t c_arg_cnt);
+
+jerry_value_t jerry_validate_and_assign_args (const jerry_value_t *js_arg_p,
+                                              const jerry_length_t js_arg_cnt,
+                                              const jerry_arg_t *c_arg_p,
+                                              const jerry_length_t c_arg_cnt);
+
+/* Helper functions for jerry_arg_iterator_t. */
+jerry_value_t jerry_arg_iterator_pop (jerry_arg_iterator_t *iterator_p);
+void *jerry_arg_iterator_get_dest (jerry_arg_iterator_t *iterator_p);
+
+/* Inline functions for initializing jerry_arg_t */
+/* The argument's type must be the primitive, e.g. number, string, boolean. */
+#define JERRY_ARG_STRICT true
+/* The validator inside will invoke toNumber, toBoolean or toString */
+#define JERRY_ARG_RELAX false
+/* The argument is optional */
+#define JERRY_ARG_OPTIOANL true
+/* The argument is required */
+#define JERRY_ARG_REQUIRED false
+static inline jerry_arg_t jerry_arg_number (double *dest, bool is_strict, bool is_optional);
+static inline jerry_arg_t jerry_arg_boolean (bool *dest, bool is_strict, bool is_optional);
+static inline jerry_arg_t jerry_arg_string (char *dest, uint32_t size, bool is_strict, bool is_optional);
```

Comment:
Yes, but firstly, we'd better have jerry_port_malloc in targets, and the malloc in jerry-libc, right?